File: native-macos/Sources/SeleneNative/Models/BangumiItem.swift
import Foundation

struct BangumiRating: Codable, Hashable {
    var total: Int
    var count: [String: Int]
    var score: Double
}

struct BangumiImages: Codable, Hashable {
    var large: String
    var common: String
    var medium: String
    var small: String
    var grid: String

    var bestImageUrl: String {
        [large, common, medium, small, grid].first { !$0.isEmpty } ?? ""
    }
}

struct BangumiCollection: Codable, Hashable {
    var doing: Int
    var onHold: Int
    var dropped: Int
    var wish: Int
    var collect: Int

    enum CodingKeys: String, CodingKey {
        case doing, dropped, wish, collect
        case onHold = "on_hold"
    }
}

struct BangumiWeekday: Codable, Hashable {
    var en: String
    var cn: String
    var ja: String
    var id: Int
}

struct BangumiItem: Identifiable, Codable, Hashable {
    var id: Int
    var url: String
    var type: Int
    var name: String
    var nameCn: String?
    var summary: String
    var airDate: String
    var airWeekday: Int
    var rating: BangumiRating
    var rank: Int
    var images: BangumiImages
    var collection: BangumiCollection

    enum CodingKeys: String, CodingKey {
        case id, url, type, name, summary, rating, rank, images, collection
        case nameCn = "name_cn"
        case airDate = "air_date"
        case airWeekday = "air_weekday"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(Int.self, forKey: .id)
        url = try container.decodeIfPresent(String.self, forKey: .url) ?? ""
        type = try container.decodeIfPresent(Int.self, forKey: .type) ?? 0
        name = Self.decodeHTMLEntities(try container.decodeIfPresent(String.self, forKey: .name) ?? "")
        nameCn = Self.decodeHTMLEntities(try container.decodeIfPresent(String.self, forKey: .nameCn) ?? "")
        summary = Self.decodeHTMLEntities(try container.decodeIfPresent(String.self, forKey: .summary) ?? "")
        airDate = try container.decodeIfPresent(String.self, forKey: .airDate) ?? ""
        airWeekday = try container.decodeIfPresent(Int.self, forKey: .airWeekday) ?? 0
        rating = try container.decodeIfPresent(BangumiRating.self, forKey: .rating) ?? BangumiRating(total: 0, count: [:], score: 0)
        rank = try container.decodeIfPresent(Int.self, forKey: .rank) ?? 0
        images = try container.decodeIfPresent(BangumiImages.self, forKey: .images) ?? BangumiImages(large: "", common: "", medium: "", small: "", grid: "")
        collection = try container.decodeIfPresent(BangumiCollection.self, forKey: .collection) ?? BangumiCollection(doing: 0, onHold: 0, dropped: 0, wish: 0, collect: 0)
    }

    static func decodeHTMLEntities(_ value: String) -> String {
        value
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
    }
}

struct BangumiDetails: Codable, Hashable {
    var id: Int
    var type: Int
    var name: String
    var nameCn: String?
    var summary: String
    var nsfw: Bool
    var locked: Bool
    var date: String?
    var platform: String?
    var images: BangumiImages
    var infobox: [String]
    var eps: Int
    var totalEpisodes: Int
    var rating: BangumiRating
    var collection: BangumiCollection
    var tags: [String]
    var metaTags: [String]
    var series: Bool
}

struct BangumiCalendarResponse: Codable {
    var weekday: BangumiWeekday
    var items: [BangumiItem]
}

File: native-windows/src/SeleneNative.Core/Models/BangumiItem.cs
using System.Text.Json.Serialization;

namespace SeleneNative.Core.Models;

public sealed class BangumiRating
{
    [JsonPropertyName("total")]
    public int Total { get; set; }

    [JsonPropertyName("count")]
    public Dictionary<string, int> Count { get; set; } = [];

    [JsonPropertyName("score")]
    public double Score { get; set; }
}

public sealed class BangumiImages
{
    [JsonPropertyName("large")]
    public string Large { get; set; } = string.Empty;

    [JsonPropertyName("common")]
    public string Common { get; set; } = string.Empty;

    [JsonPropertyName("medium")]
    public string Medium { get; set; } = string.Empty;

    [JsonPropertyName("small")]
    public string Small { get; set; } = string.Empty;

    [JsonPropertyName("grid")]
    public string Grid { get; set; } = string.Empty;

    public string BestImageUrl => new[] { Large, Common, Medium, Small, Grid }
        .FirstOrDefault(value => !string.IsNullOrWhiteSpace(value)) ?? string.Empty;
}

public sealed class BangumiCollection
{
    [JsonPropertyName("doing")]
    public int Doing { get; set; }

    [JsonPropertyName("on_hold")]
    public int OnHold { get; set; }

    [JsonPropertyName("dropped")]
    public int Dropped { get; set; }

    [JsonPropertyName("wish")]
    public int Wish { get; set; }

    [JsonPropertyName("collect")]
    public int Collect { get; set; }
}

public sealed class BangumiWeekday
{
    [JsonPropertyName("en")]
    public string En { get; set; } = string.Empty;

    [JsonPropertyName("cn")]
    public string Cn { get; set; } = string.Empty;

    [JsonPropertyName("ja")]
    public string Ja { get; set; } = string.Empty;

    [JsonPropertyName("id")]
    public int Id { get; set; }
}

public sealed class BangumiItem
{
    private string _name = string.Empty;
    private string _nameCn = string.Empty;
    private string _summary = string.Empty;

    [JsonPropertyName("id")]
    public int Id { get; set; }

    [JsonPropertyName("url")]
    public string Url { get; set; } = string.Empty;

    [JsonPropertyName("type")]
    public int Type { get; set; }

    [JsonPropertyName("name")]
    public string Name
    {
        get => _name;
        set => _name = JsonModelHelpers.DecodeHtmlEntities(value ?? string.Empty);
    }

    [JsonPropertyName("name_cn")]
    public string NameCn
    {
        get => _nameCn;
        set => _nameCn = JsonModelHelpers.DecodeHtmlEntities(value ?? string.Empty);
    }

    [JsonPropertyName("summary")]
    public string Summary
    {
        get => _summary;
        set => _summary = JsonModelHelpers.DecodeHtmlEntities(value ?? string.Empty);
    }

    [JsonPropertyName("air_date")]
    public string AirDate { get; set; } = string.Empty;

    [JsonPropertyName("air_weekday")]
    public int AirWeekday { get; set; }

    [JsonPropertyName("rating")]
    public BangumiRating Rating { get; set; } = new();

    [JsonPropertyName("rank")]
    public int Rank { get; set; }

    [JsonPropertyName("images")]
    public BangumiImages Images { get; set; } = new();

    [JsonPropertyName("collection")]
    public BangumiCollection Collection { get; set; } = new();

    public string DisplayTitle => !string.IsNullOrWhiteSpace(NameCn) ? NameCn : Name;
}

public sealed class BangumiCalendarResponse
{
    [JsonPropertyName("weekday")]
    public BangumiWeekday Weekday { get; set; } = new();

    [JsonPropertyName("items")]
    public List<BangumiItem> Items { get; set; } = [];
}
